### DISCUSSION TITLE
Versioning individual Varnomen webpages

### DISCUSSION THREAD

**Discussion by rileygr** [HGVSnomenclature/hgvs-nomenclature#12] [HGVS presentation]
It would be helpful if the Varnomen webpages were versioned beyond the publication/proposal-date versioning system currently in use.  
I sometimes resort to the Wayback Machine to see if / when the Varnomen pages change.  The community depends on the pages for the HGVS standard and additions/clarifications on the pages don’t necessarily get captured by the publication-date versioning.  From the [Versioning page](http://varnomen.hgvs.org/versioning/):

> Note the version does not change when a typing error is corrected, an example added, an explanation clarified or a question answered.

Those examples, explanations and answers can represent an effective expansion of the guidelines.

**Comment by jfjlaros** [comment]
Maybe we could think about some form of [semantic versioning](https://semver.org/), e.g.,

- Major version increments for backwards incompatible changes.
- Minor version increments for backwards compatible additions or deprecations (still accepted but no longer recommended). 
- Patch version increments for minor changes like typing errors, added examples, etc.

> I sometimes resort to the Wayback Machine to see if / when the Varnomen pages change.

The guidelines are generated from a source that is under version control. You may want to check out the [commit log](https://github.com/HGVSnomenclature/VarNomen/commits/master).

**Reply by ifokkema** [reply]
> The guidelines are generated from a source that is under version control. You may want to check out the [commit log](https://github.com/HGVSnomenclature/VarNomen/commits/master).

Actually, the correct link is [here](https://github.com/HGVSnomenclature/VarNomen/commits/gh-pages). The confusion between the branches in the Varnomen repo is due to using Github pages in combination with being maintained by a non-expert. It is very much possible (actually, highly likely) that there are conflicts between `master` and `gh-pages`. The latter is published online, but the former also sometimes receives commits.

**Comment by ifokkema** [comment]
+1. There are many changes made to the nomenclature documentation, including a change of recommendations, that did not lead to an update of any version and were not documented as a change anywhere except for in the [commit log](https://github.com/HGVSnomenclature/VarNomen/commits/gh-pages).

**Comment by ahwagner** [comment]
👍. I agree that all changes should be versioned and that versions of the HGVS nomenclature pages should be clearly indicated and browsable. For the GA4GH Variation Representation Specification (VRS), we have used semantic versioning scheme hosted on ReadTheDocs, which has allowed us to provide not only the most recent versioned draft ([stable](https://vrs.ga4gh.org/en/stable)) but also enable users to browse between other major, minor, and patch versions. These tools allow you to open a panel to browse versions in the lower-left hand corner of the spec, which it looks like this:

![image](https://user-images.githubusercontent.com/811065/212773844-8f4ccd6e-ebe8-429c-b5b1-95d837f52aad.png)

In between version tags our "cutting edge" build ([latest](https://vrs.ga4gh.org/en/latest)) provides the latest and greatest to preview new commits before versioning. I imagine future versions of HGVS could be maintained similarly.

Another cool thing this framework enables us to do is **automatically preview new changes in docs** using RTD CI. Here is an [example doc build](https://fusions--25.org.readthedocs.build/en/25/) off of an open PR for the VICC gene fusions specification. This is built and linked into an [open pull request on the fusions spec](https://github.com/cancervariants/fusions/pull/25), minimizing technical burden for content reviewers (i.e. compiling PR code on a local dev build for documentation):

![image](https://user-images.githubusercontent.com/811065/212776440-d411458b-f6ea-4111-87d1-451f278229e5.png)

I think these features would be very valuable to HVNC.

**Comment by jtdendunnen** [comment]
indeed the Nomenclature pages derive from the gh-pages branch. This 
branch is not linked to the "master"branch, i.e. the master branch is 
not maintained at all.

Johan

**Reply by jfjlaros** [reply]
Then I would suggest to remove the `main` branch.

**Reply by ifokkema** [reply]
There are 20 commits in `master` that are not in `gh-pages`, but I suppose it'll be too complex trying to unravel that. I suppose (but didn't check) that merging `master` into `gh-pages` before deleting `master` will result in a hefty conflict.

**Comment by rileygr** [comment]
It seems like this is being addressed in 
https://github.com/HGVSnomenclature/HVNC/discussions/17#discussioncomment-4783170
That's great!

**Reply by reece** [reply]
@rileygr : There are several related efforts underway regarding versioning:

- Use repository tags for version numbers. This means that the documentation will always correspond to a precise version of the repo.
- Define the rules for versions. We have discussed using rules akin to Semantic Versioning. 
- Enable switching between versions in the documentation site.

If you have any input to share about how versioning should work, this is a good time to help shape them. As @jfjlaros shared above, one of our proposed directions is like this;

* The "patch" version (z in x.y.z) is reserved for changes that fix or clarify intentions (akin to a bug fix)
* The "minor" version (y) will be used for new features that are consistent with prior versions but should not break them
* The "major" version (x) will be used when changes break existing behavior. I can only think of hypothetical cases for this, but it's possible eventually.

**Reply by rileygr** [reply]
Hi @reece!

>     * Use repository tags for version numbers. This means that the documentation will always correspond to a precise version of the repo.

If by 'repository tags' you mean 'commit' numbers for version numbers, I would suggest some instructions as to where/how to view the changes for users that are not familiar with GitHub and your repo structure.  I poked around a bit looking for documentation commits and didn't find easily them.  

>     * Define the rules for versions. We have discussed using rules akin to Semantic Versioning.

I think Semantic Versioning  as you described it is appropriate and easy to follow, as well as SV being a versioning standard.

>     * Enable switching between versions in the documentation site.

I really like the idea of being able to switch versions, whether that's in 'Read the Docs' or something else.  I'm not very familiar with read the docs, but it would be really helpful if changes were 'spotlighted' in some fashion so that minor-but-important changes are easy to find on a page.  Keep in mind that the HGVS Nomenclature  page users are frequently going to be novice-to-competent HGVS users who are not going to be easily able to identify changes without some help.

Thanks!  I really appreciate the effort you folks are putting in to improving this!  This will be a significant benefit to the community.

**Reply by reece** [reply]
> If by 'repository tags' you mean 'commit' numbers for version numbers,

No, I do not mean commit hashes. Tags are feature of git and github that essentially allow you to name (tag) a specific commit. It is common practice to use tags to associate a version like "1.2.3" with a commit hash like 3019dde1.  So, by this comment, I really mean that there's exactly one place to associate a version with a hash. 

And, yes, we should create a change log. That's a good idea and I'll add it to the list of considerations.

**Comment by rileygr** [comment]
@reece
> use tags to associate a version like "1.2.3" with a commit hash like 3019dde1

That sounds ideal!

**Comment by reece** [comment]
@rileygr Please see https://github.com/HGVSnomenclature/hgvs-nomenclature/discussions/17#discussioncomment-7603782, and especially the part about versioning. I would appreciate your comments.

**Reply by rileygr** [reply]
@reece Great job!  I'll comment on #17 , which is still open.

**Comment by reece** [comment]
I believe that this is now addressed in https://hgvs-nomenclature.org/en/latest/background/versioning/. Closing discussion.
